File: src/HelloOgre3D/sandbox/state/ShootState.cpp
#include "ShootState.h"
#include "object/AgentObject.h"
#include "GameDefine.h"

ShootState::ShootState(AgentObject* pAgent) : AgentState(pAgent)
{
	m_stateId = "ShootState";
}

ShootState::~ShootState()
{

}

void ShootState::OnEnter()
{
	m_pAgent->RequestState(SSTATE_FIRE);
}

void ShootState::OnLeave()
{

}

std::string ShootState::OnUpdate(float dt)
{
	if (m_pAgent->IsMoving())
	{
		m_pAgent->SlowMoving();
	}
	if (!m_pAgent->IsAnimReadyForShoot())
	{
		if (!m_pAgent->IsHasNextAnim())
		{
			m_pAgent->RequestState(SSTATE_FIRE);
		}
		return ""; // ¼ÌÐøµÈ´ý
	}

	auto pInput = m_pAgent->GetInput();
	if (pInput->isKeyDown(OIS::KC_2))
	{
		return "";
	}
	else if (pInput->isKeyDown(OIS::KC_3))
	{
		return "toMove";
	}
	else if (pInput->isKeyDown(OIS::KC_4))
	{
		return "toDeath";
	}

	return "toIdle";
}

File: bin/res/scripts/agent/states/ShootState.lua
-- ShootState --

--local m_pAgent = owner;
AgentState_OnEnter = function(agent)
    agent:RequestState(SSTATE_FIRE);
end

AgentState_OnLeave = function(agent)

end

AgentState_OnUpdate = function(agent, deltaTime)
    if agent:IsMoving() then
        agent:SlowMoving(2.0)
    end

    if not agent:IsAnimReadyForMove() then
        if not agent:HasNextAnim() then
            agent:RequestState(SSTATE_FIRE)
        end
        return "";
    end
    
    local pInput = agent:GetInput();
    if IsNumKeyDown(pInput, 2) then
        return "toShoot";
    elseif IsNumKeyDown(pInput, 3) then
        return "toMove";
    elseif IsNumKeyDown(pInput, 4) then
        return "toDeath";
    end

    return "";
end
